app/src/main/java/com/chouten/app/ui/views/home/HomePageViewModel.kt

```diff
@@ -35,94 +37,89 @@ class HomePageViewModel(
     val loadedModule: ModuleModel?
         get() = _loadedModule.value
 
+    var errors = 0;
+
     init {
         webview.initialize(context)
+        webview.dontCloseOnError()
+        webview.setCallback(this::callback)
     }
 
-    private val handler = Handler(Looper.getMainLooper())
+    // This is called by the webview
+    // when it's done calling the logic function
+    fun callback(message: String) {
+
+        val action = Mapper.parse<ModuleAction>(message).action;
 
-    fun initialize() {
-        viewModelScope.launch {
-            loadHomePage()
+        try {
+            if(action == "error"){
+                val error = Mapper.parse<ErrorAction>(message);
+                throw Exception(error.result);
+            }
+
+            val results = Mapper.parse<ModuleResponse<List<HomeResult>>>(message)
+            
+            _homeResults.clear()
+            _homeResults.addAll(results.result)
+        } catch (e: Exception) {
+            this.errors += 1;
+            PrimaryDataLayer.enqueueSnackbar(
+                    SnackbarVisualsWithError(
+                            e.localizedMessage ?: "Error parsing home page results.",
+                            isError = true
+                    )
+            )
+            e.printStackTrace()
         }
+
+        isLoading = false
     }
 
-    private suspend fun loadHomePage() {
-        _loadedModule.value = ModuleLayer.selectedModule
-        val homeModule = ModuleLayer.selectedModule ?: return
-        isLoading = true
-        homeModule.subtypes.forEach { subtype ->
-            // wait for homeModule code to be loaded asynchronously using run handler
-            handler.postDelayed(object : Runnable {
-                override fun run() {
-                    println("Checking if homeModule code is loaded...")
-                    if (homeModule.code?.get(subtype)?.home?.get(0)?.code != null) {
-                        println("HomeModule code is loaded!")
-                        val homeFns = homeModule.code?.get(subtype)?.home
-                        if (homeFns != null) {
-                            viewModelScope.launch {
-                                runHomeFns(homeFns)
-                            }
-                        }
-
-                        handler.removeCallbacks(this)
-                        return
-                    }
-                    handler.postDelayed(this, 100)
-                }
-            }, 100)
+    private val handler = Handler(Looper.getMainLooper())
 
+    fun initialize(shouldReset: Boolean = false) {
```

Comment:
If I'm not mistaken, the logic here can be moved into the `init` of the class itself.

Reply:
initialize is called by the HomePage. It can be refactored. I'll add a TODO.